### PR TITLE
Refactor gas estimation logic; implement gas estimator for Optimism

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

File: core/chains/configs.go [hunk 1]
```diff
@@ -168,8 +166,8 @@ func setConfigs() {
 	arbitrumMainnet.EthGasPriceDefault = *big.NewInt(1000000000000) // Arbitrum uses something like a Vickrey auction model where gas price represents a "max bid". In practice we usually pay much less
 	arbitrumMainnet.EthMaxGasPriceWei = *big.NewInt(1000000000000)  // Fix the gas price
 	arbitrumMainnet.EthMinGasPriceWei = *big.NewInt(1000000000000)  // Fix the gas price
-	arbitrumMainnet.GasUpdaterEnabled = false
-	arbitrumMainnet.GasUpdaterBlockHistorySize = 0 // Force an error if someone set GAS_UPDATER_ENABLED=true by accident; we never want to run the gas updater on arbitrum
+	arbitrumMainnet.GasEstimatorMode = "FixedPrice"
+	arbitrumMainnet.BlockHistoryEstimatorBlockHistorySize = 0 // Force an error if someone set GAS_UPDATER_ENABLED=true by accident; we never want to run the gas updater on arbitrum
```
Comment (PiotrTrzpil) [inline core/chains/configs.go:170]: GAS_UPDATER_ENABLED is no more, so the comment is now outdated

File: core/chains/configs.go [hunk 2]
```diff
@@ -183,8 +181,8 @@ func setConfigs() {
 	optimismMainnet.EthHeadTrackerHistoryDepth = 10
 	optimismMainnet.EthHeadTrackerSamplingInterval = 1 * time.Second
 	optimismMainnet.EthTxResendAfterThreshold = 15 * time.Second
-	optimismMainnet.GasUpdaterBlockHistorySize = 0 // Force an error if someone set GAS_UPDATER_ENABLED=true by accident; we never want to run the gas updater on optimism
-	optimismMainnet.GasUpdaterEnabled = false
+	optimismMainnet.BlockHistoryEstimatorBlockHistorySize = 0 // Force an error if someone set GAS_UPDATER_ENABLED=true by accident; we never want to run the gas updater on optimism
```
Comment (PiotrTrzpil) [inline core/chains/configs.go:184]: GAS_UPDATER_ENABLED is no more, so the comment is now outdated

File: core/services/bulletprooftxmanager/bulletprooftxmanager.go [hunk 3]
```diff
@@ -44,14 +48,17 @@ type Config interface {
...
 	EthNonceAutoSync() bool
 	EthRPCDefaultBatchSize() uint32
-	EthTxReaperThreshold() time.Duration
 	EthTxReaperInterval() time.Duration
+	EthTxReaperThreshold() time.Duration
 	EthTxResendAfterThreshold() time.Duration
+	GasEstimatorMode() string
 	TriggerFallbackDBPollInterval() time.Duration
-	EthMaxInFlightTransactions() uint32
+	// TODO: Remove this optimism gas fees thing
```
Comment (PiotrTrzpil) [inline core/services/bulletprooftxmanager/bulletprooftxmanager.go:61]: could be a task?

File: core/services/bulletprooftxmanager/bulletprooftxmanager.go [hunk 4]
```diff
@@ -101,6 +108,7 @@ type BulletproofTxManager struct {
 	keyStore         KeyStore
 	advisoryLocker   postgres.AdvisoryLocker
 	eventBroadcaster postgres.EventBroadcaster
+	estimator        gas.Estimator
```
Comment (PiotrTrzpil) [inline core/services/bulletprooftxmanager/bulletprooftxmanager.go:111]: hm minor, but `gasEstimator` seems like more readable field name?

File: core/services/bulletprooftxmanager/bulletprooftxmanager.go [hunk 5]
```diff
@@ -337,30 +354,12 @@ func SendEther(db *gorm.DB, from, to common.Address, value assets.Eth, gasLimit
 	return etx, err
 }
 
-func newAttempt(ctx context.Context, ethClient eth.Client, ks KeyStore, config Config, etx models.EthTx, suggestedGasPrice *big.Int) (models.EthTxAttempt, error) {
+// TODO: Remove gas limit multiplier?
```
Comment (PiotrTrzpil) [inline core/services/bulletprooftxmanager/bulletprooftxmanager.go:357]: the comment should be elsewhere? I can't see 'gas limit multiplier' in this function

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 6]
```diff
@@ -455,7 +463,10 @@ func saveAttempt(db *gorm.DB, etx *models.EthTx, attempt models.EthTxAttempt, ne
 }
 
 func (eb *EthBroadcaster) tryAgainWithHigherGasPrice(sendError *eth.SendError, etx models.EthTx, attempt models.EthTxAttempt, initialBroadcastAt time.Time) error {
-	bumpedGasPrice, err := BumpGas(eb.config, attempt.GasPrice.ToInt())
+	// bumpedGasPrice, err := BumpGas(eb.config, attempt.GasPrice.ToInt())
```
Comment (PiotrTrzpil) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:466]: to remove?

File: core/services/gas/block_history_estimator.go [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+// in and we are not currently fetching
+func (b *blockHistoryEstimator) OnNewLongestChain(ctx context.Context, head models.Head) {
+	b.mb.Deliver(head)
+}
+
+func (b *blockHistoryEstimator) Start() error {
+	return b.StartOnce("BlockHistoryEstimator", func() error {
+		b.logger.Debugw("BlockHistoryEstimator: starting")
+		if uint(b.config.BlockHistoryEstimatorBlockHistorySize()) > b.config.EthFinalityDepth() {
+			b.logger.Warnf("BlockHistoryEstimator: GAS_UPDATER_BLOCK_HISTORY_SIZE=%v is greater than ETH_FINALITY_DEPTH=%v, blocks deeper than finality depth will be refetched on every gas updater cycle, causing unnecessary load on the eth node. Consider decreasing GAS_UPDATER_BLOCK_HISTORY_SIZE or increasing ETH_FINALITY_DEPTH", b.config.BlockHistoryEstimatorBlockHistorySize(), b.config.EthFinalityDepth())
```
Comment (PiotrTrzpil) [inline core/services/gas/block_history_estimator.go:106]: should use new env variable names

File: core/services/gas/block_history_estimator.go [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+	blockDelay := int64(b.config.BlockHistoryEstimatorBlockDelay())
+	historySize := int64(b.config.BlockHistoryEstimatorBlockHistorySize())
+
+	if historySize <= 0 {
+		return errors.Errorf("BlockHistoryEstimator: history size must be > 0, got: %d", historySize)
+	}
+
+	highestBlockToFetch := head.Number - blockDelay
+	if highestBlockToFetch < 0 {
+		return errors.Errorf("BlockHistoryEstimator: cannot fetch, current block height %v is lower than GAS_UPDATER_BLOCK_DELAY=%v", head.Number, blockDelay)
```
Comment (PiotrTrzpil) [inline core/services/gas/block_history_estimator.go:236]: env name to update

File: core/services/gas/models.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	} else if bumpedGasPrice.Cmp(originalGasPrice) == 0 {
+		// NOTE: This really shouldn't happen since we enforce minimums for
+		// ETH_GAS_BUMP_PERCENT and ETH_GAS_BUMP_WEI in the config validation,
+		// but it's here anyway for a "belts and braces" approach
+		return bumpedGasPrice, errors.Errorf("bumped gas price of %s is equal to original gas price of %s."+
+			" ACTION REQUIRED: This is a configuration error, you must increase either "+
+			"ETH_GAS_BUMP_PERCENT or ETH_GAS_BUMP_WEI", bumpedGasPrice.String(), originalGasPrice.String())
+	}
+	// TODO: Move/fix these
+	// promNumGasBumps.Inc()
```
Comment (PiotrTrzpil) [inline core/services/gas/models.go:204]: could be moved to the block history estimator?

Comment (samsondav) [reply core/services/gas/models.go:212]: Decided to keep it here to maintain compatibility

File: core/services/bulletprooftxmanager/eth_confirmer_test.go [hunk 10]
```diff
@@ -1962,7 +1961,8 @@ func TestEthConfirmer_EnsureConfirmedTransactionsInLongestChain(t *testing.T) {
 	config, cleanup := cltest.NewConfig(t)
 	defer cleanup()
 
-	ec := bulletprooftxmanager.NewEthConfirmer(store.DB, ethClient, config, ethKeyStore, &postgres.NullAdvisoryLocker{}, []ethkey.Key{key})
+	// ec := bulletprooftxmanager.NewEthConfirmer(store.DB, ethClient, config, ethKeyStore, &postgres.NullAdvisoryLocker{}, []ethkey.Key{key}, nil)
```
Comment (PiotrTrzpil) [inline core/services/bulletprooftxmanager/eth_confirmer_test.go:1964]: to remove?

File: core/services/eth/errors.go [hunk 11]
```diff
@@ -148,6 +156,18 @@ func (s *SendError) IsTooExpensive() bool {
 	return s.is(TooExpensive)
 }
 
+// IsFeeTooLow is an optimism-specific error returned when total fee is too low
+// TODO: Add test
+// TODO: Add is fee too high error - see https://www.notion.so/How-to-pay-Fees-in-Optimistic-Ethereum-f706f4e5b13e460fa5671af48ce9a695
```
Comment (PiotrTrzpil) [inline core/services/eth/errors.go:161]: probably outdated? as it's implemented by IsFeeTooHigh methd

File: core/services/eth/errors.go [hunk 12]
```diff
@@ -93,7 +95,13 @@ var arbitrum = ClientErrors{
 	Fatal:                 arbitrumFatal,
 }
 
-var clients = []ClientErrors{parity, geth, arbitrum}
+var optimism = ClientErrors{
+	// TODO: Test these
```
Comment (PiotrTrzpil) [inline core/services/eth/errors.go:99]: might be worth creating a task for this one

File: core/services/gas/block_history_estimator.go [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+		gasPrice   *big.Int
+		gasPriceMu sync.RWMutex
+
+		logger *logger.Logger
+	}
+)
+
+// NewBlockHistoryEstimator returns a new BlockHistoryEstimator that listens
+// for new heads and updates the base gas price dynamically based on the
+// confibred percentile of gas prices in that block
```
Comment (PiotrTrzpil) [inline core/services/gas/block_history_estimator.go:72]: typo: confibred

File: core/services/gas/block_history_estimator.go [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+		ethClient,
+		config,
+		make([]Block, 0),
+		utils.NewMailbox(1),
+		new(sync.WaitGroup),
+		ctx,
+		cancel,
+		nil,
+		sync.RWMutex{},
+		logger.CreateLogger(logger.Default.With("id", "gas_updater")),
```
Comment (PiotrTrzpil) [inline core/services/gas/block_history_estimator.go:86]: hmm do we want to change the logger id as well?

File: core/services/gas/block_history_estimator.go [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+		"headNum", head.Number,
+		"blocks", numsInHistory,
+	)
+	b.setPercentileGasPrice(percentileGasPrice)
+	promBlockHistoryEstimatorSetGasPrice.WithLabelValues(fmt.Sprintf("%v%%", percentile)).Set(float64(percentileGasPrice.Int64()))
+}
+
+func (b *BlockHistoryEstimator) FetchBlocks(ctx context.Context, head models.Head) error {
+	// HACK: blockDelay is the number of blocks that the gas updater trails behind head.
+	// E.g. if this is set to 3, and we receive block 10, gas updater will
```
Comment (PiotrTrzpil) [inline core/services/gas/block_history_estimator.go:222]: could update the 'gas updater' parts of the comment

File: core/services/gas/optimism_estimator.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+		return err
+	}
+	g.L1GasPrice = l1.ToInt()
+	g.L2GasPrice = l2.ToInt()
+	return nil
+}
+
+func (o *optimismEstimator) refreshPrices() (t *time.Timer) {
+	var res OptimismGasPricesResponse
+	t = time.NewTimer(utils.WithJitter(o.pollPeriod))
```
Comment (PiotrTrzpil) [inline core/services/gas/optimism_estimator.go:121]: hm something seems to be missing with the usage of this timer - it's just returned from this method, but then does not seem to be actually used outside either?

File: core/services/gas/optimism_estimator.go [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+	config     Config
+	client     optimismRPCClient
+	pollPeriod time.Duration
+
+	gasPriceMu sync.RWMutex
+	l1GasPrice *big.Int
+	l2GasPrice *big.Int
+
+	chForceRefetch chan (chan struct{})
```
Comment (PiotrTrzpil) [inline core/services/gas/optimism_estimator.go:42]: good to know about this double-channel technique for a request and response 👍🏼

File: core/services/eth/errors.go (hunk 12 above)
Comment (samsondav) [reply core/services/eth/errors.go:99]: I just added the test to this pr

File: core/utils/backoff_ticker_test.go [hunk 18]
```diff
@@ -2,50 +2,46 @@ package utils_test
...
+	// assertTick(t, &ticker, false, 200*time.Millisecond)
 
-	ticker.Start()
-	assertTick(t, &ticker, true, 50*time.Millisecond)
-	// Does not tick before the next backoff period
-	assertTick(t, &ticker, false, 50*time.Millisecond)
+	// ticker.Start()
+	// assertTick(t, &ticker, true, 50*time.Millisecond)
+	// // Does not tick before the next backoff period
+	// assertTick(t, &ticker, false, 50*time.Millisecond)
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/utils/backoff_ticker_test.go:46]: I actually think we should remove this test

File: core/services/gas/optimism_estimator.go (hunk 16 above)
Comment (samsondav) [reply core/services/gas/optimism_estimator.go:121]: Yup, it is used in `run()`

Comment (PiotrTrzpil) [reply core/services/gas/optimism_estimator.go:120]: ah yes `case <-t.C:`

**Review by PiotrTrzpil — DISMISSED** [review]

**Review by PiotrTrzpil — APPROVED** [review]

**Pull request merged**